### PR TITLE
Design And Initial Implementation of Default Function Registry

### PR DESCRIPTION
**Pull request by VladislavsPerkanuks**
Initial implementation design for integrating functions (default and custom) into the template:

Currently, only two functions have been included:

- `string`
- `number` (though not fully implemented due to complexity in accommodating all options

### CODE REVIEW

File: template/registry/registry.go [hunk 1]
```diff
@@ -0,0 +1,153 @@
+package registry
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+)
+
+type Registry []Func
```
Comment (janishorsts) [inline template/registry/registry.go:9]: Why not `map[string]Func`? It speeds up Func lookup.

File: template/registry/registry.go [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+	for _, fn := range r {
+		if fn.Name == name {
+			return fn, true
+		}
+	}
+
+	return Func{}, false
+}
+
+func (r *Func) Format(input any, options map[string]any) (any, error) {
```
Comment (janishorsts) [inline template/registry/registry.go:56]: ? 

```suggestion
func (f *Func) Format(input any, options map[string]any) (any, error) {
```

File: template/registry/registry.go [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+	}
+
+	if err := r.FormatSignature.check(input, options); err != nil {
+		return "", fmt.Errorf("check input: %w", err)
+	}
+
+	return r.F(input, options)
+}
+
+func (r *Func) Match(input any, options map[string]any) (any, error) {
```
Comment (janishorsts) [inline template/registry/registry.go:68]: ?

```suggestion
func (f *Func) Match(input any, options map[string]any) (any, error) {
```

File: template/registry/registry.go [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+	if err := r.MatchSignature.check(input, options); err != nil {
+		return "", fmt.Errorf("check input: %w", err)
+	}
+
+	return r.F(input, options)
+}
+
+func (s *Signature) check(input any, options map[string]any) error {
+	if s.Input && input == nil {
+		return errors.New("input is nil, but required")
```
Comment (janishorsts) [inline template/registry/registry.go:82]: Alternatively

```suggestion
		return errors.New("input is required, got nil")
```

File: template/registry/registry.go [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+	}
+
+	if !s.Input && input != nil {
+		return errors.New("input is not allowed")
+	}
+
+	if s.Options == nil && len(options) > 0 {
```
Comment (janishorsts) [inline template/registry/registry.go:89]: Is it supposed to allow an empty slice in `s.Options`?

```suggestion
	if len(s.Options) == 0 && len(options) > 0 {
```

Comment (VladislavsPerkanuks) [reply template/registry/registry.go:89]: nope, u right

File: template/registry/registry.go (hunk 1 above)
Comment (VladislavsPerkanuks) [reply template/registry/registry.go:9]: dunno :)

File: template/registry/registry.go [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+func (o Options) check(got map[string]any) error {
+	for name, val := range got {
+		opt, ok := o.find(name)
+		if !ok {
+			validOptions := make([]string, 0, len(o))
+			for _, opt := range o {
+				validOptions = append(validOptions, opt.Name)
+			}
+
+			return fmt.Errorf("unknown option: '%s'. Valid options: [%s]", name, strings.Join(validOptions, ", "))
```
Comment (janishorsts) [inline template/registry/registry.go:117]: Alternatively

```suggestion
			return fmt.Errorf("expected one of options - %s, got %s", name, strings.Join(validOptions, ", "))
```

File: template/registry/registry.go [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+		}
+
+		if opt.ValidateValue != nil {
+			if err := opt.ValidateValue(val); err != nil {
+				return fmt.Errorf("validate options value '%s': %w", name, err)
```
Comment (janishorsts) [inline template/registry/registry.go:122]:
```suggestion
				return fmt.Errorf("validate option value '%s': %w", name, err)
```

File: template/registry/registry.go [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+				if possible == val {
+					found = true
+					break
+				}
+			}
+
+			if !found {
+				return fmt.Errorf("invalid value '%s' for option '%s'. Valid values: %s  ", val, name, opt.PossibleValues)
+			}
+		}
```
Comment (janishorsts) [inline template/registry/registry.go:139]: Does our golang version support `slices.Contains` in this project?

```suggestion
		if !slices.Contains(opt.PossibleValues, val) {
			return fmt.Errorf("expected one of option '%s' allowed values - %s, got %s ", name, opt.PossibleValues, val)
		}
```

Comment (VladislavsPerkanuks) [reply template/registry/registry.go:139]: did not want to install additonal pkg, for one time :)

File: template/registry/registry.go [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+	"strings"
+)
+
+type Registry []Func
+
+// Func is a function that can be used in formatting and matching contexts.
+type Func struct {
+	FormatSignature *Signature                             // Signature of the function when called in formatting context
+	MatchSignature  *Signature                             // Signature of the function when called in matching context
+	F               func(any, map[string]any) (any, error) // Function itself
```
Comment (janishorsts) [inline template/registry/registry.go:15]: `Func` or `Fn` instead of `F`?

File: template/registry/number.go [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+			{
+				Name: "currency",
+				Description: `The currency to use in currency formatting.
+Possible values are the ISO 4217 currency codes, such as "USD" for the US dollar,
+"EUR" for the euro, or "CNY" for the Chinese RMB — see the
+Current currency &amp; funds code list
+(https://www.unicode.org/cldr/charts/latest/supplemental/detailed_territory_currency_information.html).
+There is no default value; if the style is "currency", the currency property must be provided.`,
+				ValidateValue: func(a any) error {
+					_, err := castAs[currency.Unit](a)
```
Comment (janishorsts) [inline template/registry/number.go:43]: This does not validate currency value, right?

Comment (VladislavsPerkanuks) [reply template/registry/number.go:43]: in this context: 
only `currency.Unit` can be passed as option value

Comment (VladislavsPerkanuks) [reply template/registry/number.go:43]: so no casting

Comment (VladislavsPerkanuks) [reply template/registry/number.go:43]: i mean, there is nothing to validate, if argument is currency.Unit, no ? 🤔

Comment (VladislavsPerkanuks) [reply template/registry/number.go:43]: i think alternative is better
```go
if _, ok := a.(currency.Unit); !ok {
return ...
}

return nil
```

Comment (janishorsts) [reply template/registry/number.go:43]: Mostly it seems to be ok, `currency.Unit` does not expose its value -  https://cs.opensource.google/go/x/text/+/refs/tags/v0.14.0:currency/currency.go;l=69

Effectively, consumers of the currency package cannot change the index directly.

However, zero value `currency.Unit` might cause an issue.

Comment (VladislavsPerkanuks) [reply template/registry/number.go:43]: yeah, it can be a problem

File: template/registry/registry.go (hunk 8 above)
Comment (VladislavsPerkanuks) [reply template/registry/registry.go:139]: wont work 

> any to satisfy comparable requires go1.20 or later

File: template/registry/registry.go [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+	"strings"
+)
+
+type Registry map[string]Func
+
+// Func is a function that can be used in formatting and matching contexts.
+type Func struct {
+	FormatSignature *Signature                             // Signature of the function when called in formatting context
+	MatchSignature  *Signature                             // Signature of the function when called in matching context
+	Fn              func(any, map[string]any) (any, error) // Function itself
```
Comment (kaaposc) [inline template/registry/registry.go:15]: Is there a case when first returned value is other than string?

Comment (VladislavsPerkanuks) [reply template/registry/registry.go:15]: it could be anything :), thats the problem.

**Review by kaaposc — APPROVED** [review]

File: template/registry/registry.go [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+	Default     any
+
+	ValidateValue  func(any) error // If option value is not restricted by a set of values.
+	PossibleValues []any           // If option value is restricted by a set of values.
+}
+
+type Options []Option
+
+// NewRegistry returns a new registry with default functions.
+func NewRegistry() Registry {
```
Comment (janishorsts) [inline template/registry/registry.go:40]: Consumers use `registry.New()`. This package unlikely to have other `New` functions.

```suggestion
func New() Registry {
```

File: template/registry/number.go [hunk 13]
```diff
@@ -0,0 +1,240 @@
...
+}
+
+// TODO: supports only style and signDisplay options.
+func numberF(input any, options map[string]any) (any, error) {
+	num, err := castAs[float64](input)
+	if err != nil {
+		return nil, fmt.Errorf("convert input to float64: %w", err)
+	}
+
+	if options == nil {
```
Comment (janishorsts) [inline template/registry/number.go:161]: Check `len` instead?

```suggestion
	if len(options) == 0 {
```

File: template/registry/number.go [hunk 14]
```diff
@@ -0,0 +1,240 @@
...
+	style, ok := options["style"]
+	if !ok {
+		style = "decimal"
+	}
+
+	switch style {
+	case "decimal":
+		result = strconv.FormatFloat(num, 'f', -1, 64)
+	case "percent":
+		result = fmt.Sprintf("%.2f%%", num*100) //nolint:gomnd
```
Comment (janishorsts) [inline template/registry/number.go:185]: Does the spec imply to use 2 decimal points by default?

Comment (VladislavsPerkanuks) [reply template/registry/number.go:185]: for percentage yes

File: template/registry/registry.go [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+			return fmt.Errorf("validate input: %w", err)
+		}
+	}
+
+	if len(options) > 0 {
+		if err := s.Options.check(options); err != nil {
+			return fmt.Errorf("check options: %w", err)
+		}
+	}
+
```
Comment (janishorsts) [inline template/registry/registry.go:95]: Let the `check()` handle empty options: keeps the flow nice and tidy.

```suggestion
	if err := s.Options.check(options); err != nil {
		return fmt.Errorf("check options: %w", err)
	}
```

Comment (janishorsts) [reply template/registry/registry.go:95]: `check` already handles empty options nicely

**Pull request merged**